Removed Serializer by restructuring. Big refactor.

vdka opened the pull request:
Serializing is now done by calling `try json.jsonString()` Cleaned up naming
- `JSONValue` -> `JSON`
- `JSON*` -> `*`
- Lowercase enum cases to avoid type name collisions and follow swift 3 guidelines
- Removed `JSONNumberType`. JSON now has two cases that cover it. `.integer` `.double`

Parser is unchanged save `JSONParser.parse[String|Data]` both are now just `JSONParser.parse`

Accessing JSON values now looks neater thanks to an extension on Optional. This extension allows a access style that is very similar to [SwiftyJSON](https://github.com/SwiftyJSON/SwiftyJSON)

Some things that are outstanding are updates to `README.md` and a new implementation of a pretty printer. The pretty printer must take a new approach now, I believe the only way for it to work now is to insert indentation into the string generated by `json.jsonString()`.

The performance comparison should also be done for parsing _and_ serialization.

Once you have reviewed the changes made I will update the `README` with new usage instructions.

vdka committed:
Shared the Example scheme fixing Carthage build

vdka committed:
Removed Serializer by restructuring.

Serializing is now done by calling `try json.jsonString()`
Cleaned up naming
- `JSONValue` -> `JSON`
- `JSON*` -> `*`
- Lowercase enum cases to avoid type name collisions and follow swift 3 guidelines
- Removed `JSONNumberType`. JSON now has two cases that cover it. `.integer` `.double`

Accessing JSON values now looks neater thanks to an extension on Optional.
This extension allows a access style that is very similar to [SwiftyJSON](https://github.com/SwiftyJSON/SwiftyJSON)

tyrone-sudeium commented:
Nice one, the diff is too large for GitHub to render so I'll check it out tonight!

tyrone-sudeium commented:
Haven't done a full review yet but some points. These are ones you should address before I'll merge:
- As per the [Swift Programming Language](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Enumerations.html#//apple_ref/doc/uid/TP40014097-CH12-ID145) book, enumeration cases should be TitleCase. I don't think I've ever seen all lower case enumeration cases before, and it's confusing because they look like method invocations.
- Something failed on Travis and I'm not sure why. Seems to be if you run the unit tests on the OS X SDK?

These I can correct myself after merging, hope you don't take offense, I just like to keep things consistent!
- Use four spaces instead of tabs for indenting. An example of why I'm so strict about this can be seen on [Line 38](https://github.com/VDKA/JSONCore/blob/5db6dca570b119c300bf3f17f34a2e09a0ecd4fb/Source/JSONCore.swift#L38).
- Some multi-line documentation comments are using `///` instead of `/**` and some aren't. All multi-line documentation comments shall be `/**`.

I will run the performance test on my Mac at home tonight where I have baselines set up to see if we've made any gains or regressions!

Thanks again for your incredible work.

vdka commented:
Regarding the failed CI there are two failed tests cases both for pretty printing.

Regarding the lowercase enum cases Swift Evolution [Proposal 0005](https://github.com/apple/swift-evolution/blob/master/proposals/0005-objective-c-name-translation.md) has been accepted and contains the following:

> Lowercase values: The Swift API Design Guidelines have non-type declarations lowercased. Lowercase non-prefixed values whenever they are imported, including enumerators (whether they end up in Swift as enum cases or option sets) and any other properties/functions (e.g., a property named URLHandler will be lowercased to urlHandler).

I don't really like the loss of TitleCase in enums but at least in this use case, it allows the cases like `.string` to exist. (as `.String`) they error out.

I will fix up indentation in my next commit, also adding a tiny bit of functionality to bring subscripting syntax in line with SwiftyJSON.

vdka committed:
Tweaked access syntax, now inline with SwiftyJSON

`json["key"]?.string` can now be `json["key"].string`

Fixed indentation and multiline comments

tyrone-sudeium commented:
Normally I'd argue that, since we're writing Swift 2.2 here and not Swift 3, you should probably use the standards for Swift 2.2 (which have title case) -- however, `.string` vs `.String`... probably not worth breaking this when we'd have to rename everything in a few months anyway.

vdka committed:
Fixed swift 2.2 debug literals `#file` & `#line`

Added setter to `Optional<_JSONType>` subscript

vdka committed:
StaticString got left behind.

This change will probably need to be changed back in 2 weeks with Swift 2.2

vdka commented:
Regarding pretty printing it looks like it may be partially implemented within a similar system to JSONCore [here](https://gist.github.com/DeFrenZ/3ddc0bbb2ae3d2b79d74)

tyrone-sudeium commented:
I've run the performance tests and we've lost a small amount of performance, but within the margin of error (10.4 seconds before, 10.7 seconds after).
